File: frontend/src/api/company.ts
import { apiClient } from './client'
import type { Company, FiscalYearStart } from '../auth/AuthContext'

export type CompanyFormValues = {
  name: string
  tradeName: string
  iceNumber: string
  rcNumber: string
  cnssNumber: string
  sector: string
  address: string
  city: string
  postalCode: string
  phone: string
  email: string
  website: string
  employeesCount: string
  fiscalYearStart: FiscalYearStart
}

export const emptyCompanyForm: CompanyFormValues = {
  name: '',
  tradeName: '',
  iceNumber: '',
  rcNumber: '',
  cnssNumber: '',
  sector: '',
  address: '',
  city: '',
  postalCode: '',
  phone: '',
  email: '',
  website: '',
  employeesCount: '',
  fiscalYearStart: 'JANUARY',
}

export function companyToForm(company: Company): CompanyFormValues {
  return {
    name: company.name,
    tradeName: company.tradeName ?? '',
    iceNumber: company.iceNumber,
    rcNumber: company.rcNumber,
    cnssNumber: company.cnssNumber,
    sector: company.sector ?? '',
    address: company.address,
    city: company.city,
    postalCode: company.postalCode ?? '',
    phone: company.phone,
    email: company.email,
    website: company.website ?? '',
    employeesCount: company.employeesCount?.toString() ?? '',
    fiscalYearStart: company.fiscalYearStart,
  }
}

function toPayload(values: CompanyFormValues) {
  const employeesCountStr = values.employeesCount.trim()
  return {
    name: values.name.trim(),
    tradeName: values.tradeName.trim() || null,
    iceNumber: values.iceNumber.trim(),
    rcNumber: values.rcNumber.trim(),
    cnssNumber: values.cnssNumber.trim(),
    sector: values.sector.trim() || null,
    address: values.address.trim(),
    city: values.city.trim(),
    postalCode: values.postalCode.trim() || null,
    phone: values.phone.trim(),
    email: values.email.trim(),
    website: values.website.trim() || null,
    employeesCount: employeesCountStr === '' ? null : Number(employeesCountStr),
    fiscalYearStart: values.fiscalYearStart,
  }
}

export async function createCompany(values: CompanyFormValues): Promise<Company> {
  const res = await apiClient.post<Company>('/companies', toPayload(values))
  return res.data
}

export async function updateCompany(values: CompanyFormValues): Promise<Company> {
  const res = await apiClient.put<Company>('/companies/me', toPayload(values))
  return res.data
}

File: backend/src/main/java/com/elguennouni/mohassib/entity/Company.java
package com.elguennouni.mohassib.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "companies", indexes = {
        @Index(name = "idx_companies_user_id", columnList = "user_id", unique = true),
        @Index(name = "idx_companies_ice_number", columnList = "ice_number"),
        @Index(name = "idx_companies_rc_number", columnList = "rc_number")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Company {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id", nullable = false, unique = true)
    private Long userId;

    @Column(nullable = false, length = 255)
    private String name;

    @Column(name = "trade_name", length = 255)
    private String tradeName;

    @Column(name = "ice_number", nullable = false, length = 15)
    private String iceNumber;

    @Column(name = "rc_number", nullable = false, length = 50)
    private String rcNumber;

    @Column(name = "cnss_number", nullable = false, length = 50)
    private String cnssNumber;

    @Column(length = 100)
    private String sector;

    @Column(nullable = false, length = 2000)
    private String address;

    @Column(nullable = false, length = 100)
    private String city;

    @Column(name = "postal_code", length = 20)
    private String postalCode;

    @Column(nullable = false, length = 20)
    private String phone;

    @Column(nullable = false, length = 255)
    private String email;

    @Column(length = 255)
    private String website;

    @Column(name = "employees_count")
    private Integer employeesCount;

    @Enumerated(EnumType.STRING)
    @Column(name = "fiscal_year_start", nullable = false, length = 20)
    private FiscalYearStart fiscalYearStart;

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    @UpdateTimestamp
    private LocalDateTime updatedAt;
}
